File: worker/src/mcp/atlassianClient.ts
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";
import type { AppConfig } from "../config.ts";
import type { IssueLike } from "../aggregate/workload.ts";

function basicAuthHeader(email: string, apiKey: string): string {
  return `Basic ${Buffer.from(`${email}:${apiKey}`).toString("base64")}`;
}

function textFromToolResult(result: {
  content?: Array<{ type: string; text?: string }>;
  structuredContent?: unknown;
}): unknown {
  if (result.structuredContent != null) return result.structuredContent;
  const texts = (result.content ?? [])
    .filter((c) => c.type === "text" && c.text)
    .map((c) => c.text as string);
  if (texts.length === 0) return null;
  const joined = texts.join("\n");
  try {
    return JSON.parse(joined);
  } catch {
    return joined;
  }
}

export type McpIssueFetchResult = {
  source: "mcp";
  issues: IssueLike[];
  tools: string[];
};

export async function fetchIssuesViaMcp(
  config: AppConfig,
): Promise<McpIssueFetchResult | { source: "mcp-unavailable"; tools: string[]; reason: string }> {
  const transport = new StreamableHTTPClientTransport(
    new URL(config.ATLASSIAN_MCP_URL),
    {
      requestInit: {
        headers: {
          Authorization: basicAuthHeader(config.JIRA_EMAIL, config.JIRA_API_KEY),
        },
      },
    },
  );

  const client = new Client({
    name: "jira-workload-worker",
    version: "0.1.0",
  });

  try {
    await client.connect(transport);
    const listed = await client.listTools();
    const tools = listed.tools.map((t) => t.name);

    if (!tools.includes("searchJiraIssuesUsingJql")) {
      return {
        source: "mcp-unavailable",
        tools,
        reason:
          "MCP session has no searchJiraIssuesUsingJql tool. Create an API token with Jira read/search scopes, and ensure Rovo MCP API-token auth is enabled for your org.",
      };
    }

    let cloudId: string | undefined;
    if (tools.includes("getAccessibleAtlassianResources")) {
      const resources = await client.callTool({
        name: "getAccessibleAtlassianResources",
        arguments: {},
      });
      const parsed = textFromToolResult(
        resources as {
          content?: Array<{ type: string; text?: string }>;
          structuredContent?: unknown;
        },
      );
      cloudId = extractCloudId(parsed, config.JIRA_SITE);
    }
    if (!cloudId) {
      cloudId = await resolveCloudIdFromSite(config.JIRA_SITE);
    }
    if (!cloudId) {
      return {
        source: "mcp-unavailable",
        tools,
        reason:
          "Could not resolve Atlassian cloudId for MCP Jira tools. Check JIRA_SITE.",
      };
    }

    const jql = `project = ${config.JIRA_PROJECT_KEY} ORDER BY updated DESC`;
    const args: Record<string, unknown> = {
      cloudId,
      jql,
      maxResults: 100,
      fields: ["summary", "status", "assignee"],
    };

    const all: IssueLike[] = [];
    const seen = new Set<string>();
    let nextPageToken: string | undefined;
    let startAt = 0;

    for (let page = 0; page < 20; page += 1) {
      const pageArgs: Record<string, unknown> = { ...args };
      if (nextPageToken) pageArgs.nextPageToken = nextPageToken;
      else if (startAt > 0) pageArgs.startAt = startAt;

      const result = await client.callTool({
        name: "searchJiraIssuesUsingJql",
        arguments: pageArgs,
      });
      if ("isError" in result && result.isError) {
        const errText = textFromToolResult(
          result as {
            content?: Array<{ type: string; text?: string }>;
            structuredContent?: unknown;
          },
        );
        return {
          source: "mcp-unavailable",
          tools,
          reason: `MCP searchJiraIssuesUsingJql failed: ${String(errText).slice(0, 300)}`,
        };
      }
      const payload = textFromToolResult(
        result as {
          content?: Array<{ type: string; text?: string }>;
          structuredContent?: unknown;
        },
      );
      const pageIssues = normalizeMcpIssues(payload);
      let added = 0;
      for (const issue of pageIssues.issues) {
        if (!issue.key || seen.has(issue.key)) continue;
        seen.add(issue.key);
        all.push(issue);
        added += 1;
      }

      if (added === 0) break;
      if (pageIssues.nextPageToken) {
        nextPageToken = pageIssues.nextPageToken;
        continue;
      }
      if (!pageIssues.hasMore) break;
      startAt += pageIssues.issues.length;
    }

    return { source: "mcp", issues: all, tools };
  } finally {
    await client.close().catch(() => undefined);
  }
}

function extractCloudId(parsed: unknown, siteUrl: string): string | undefined {
  const host = new URL(siteUrl).hostname;
  if (Array.isArray(parsed)) {
    for (const item of parsed) {
      if (item && typeof item === "object") {
        const rec = item as Record<string, unknown>;
        const url = String(rec.url ?? rec.siteUrl ?? "");
        if (url.includes(host) && typeof rec.id === "string") return rec.id;
        if (url.includes(host) && typeof rec.cloudId === "string")
          return rec.cloudId;
      }
    }
  }
  if (parsed && typeof parsed === "object") {
    const rec = parsed as Record<string, unknown>;
    if (Array.isArray(rec.resources)) return extractCloudId(rec.resources, siteUrl);
    if (typeof rec.cloudId === "string") return rec.cloudId;
  }
  return undefined;
}

async function resolveCloudIdFromSite(siteUrl: string): Promise<string | undefined> {
  try {
    const res = await fetch(
      `${siteUrl.replace(/\/$/, "")}/_edge/tenant_info`,
    );
    if (!res.ok) return undefined;
    const data = (await res.json()) as { cloudId?: string };
    return data.cloudId;
  } catch {
    return undefined;
  }
}

function normalizeMcpIssues(payload: unknown): {
  issues: IssueLike[];
  hasMore: boolean;
  nextPageToken?: string;
} {
  if (!payload) return { issues: [], hasMore: false };

  let issuesRaw: unknown[] = [];
  let total = 0;
  let startAt = 0;
  let maxResults = 0;
  let nextPageToken: string | undefined;
  let isLast: boolean | undefined;

  if (Array.isArray(payload)) {
    issuesRaw = payload;
  } else if (typeof payload === "object") {
    const rec = payload as Record<string, unknown>;
    if (Array.isArray(rec.issues)) issuesRaw = rec.issues;
    else if (Array.isArray(rec.values)) issuesRaw = rec.values;
    total = Number(rec.total ?? 0);
    startAt = Number(rec.startAt ?? 0);
    maxResults = Number(rec.maxResults ?? 0);
    if (typeof rec.nextPageToken === "string" && rec.nextPageToken) {
      nextPageToken = rec.nextPageToken;
    }
    if (typeof rec.isLast === "boolean") isLast = rec.isLast;
  }

  const issues: IssueLike[] = issuesRaw.map((raw) => {
    const issue = raw as Record<string, unknown>;
    const fields = (issue.fields ?? issue) as Record<string, unknown>;
    const status = (fields.status ?? {}) as Record<string, unknown>;
    const statusCategory = (status.statusCategory ?? {}) as Record<
      string,
      unknown
    >;
    const assignee = (fields.assignee ?? null) as Record<string, unknown> | null;
    return {
      key: String(issue.key ?? fields.key ?? ""),
      statusName: String(status.name ?? "Unknown"),
      statusCategory:
        (statusCategory.name as string | undefined) ??
        (statusCategory.key as string | undefined) ??
        null,
      assigneeName:
        (assignee?.displayName as string | undefined) ??
        (assignee?.name as string | undefined) ??
        null,
    };
  });

  if (isLast === true) {
    return { issues, hasMore: false, nextPageToken: undefined };
  }
  if (nextPageToken) {
    return { issues, hasMore: true, nextPageToken };
  }

  const hasMore =
    total > 0
      ? startAt + issues.length < total
      : maxResults > 0 && issues.length >= maxResults;

  return { issues, hasMore };
}

File: worker/app/mcp/atlassian_client.py
from __future__ import annotations

import base64
import json
from typing import Any
from urllib.parse import urlparse

import httpx
from mcp import ClientSession
from mcp.client.streamable_http import create_mcp_http_client, streamable_http_client
from mcp.types import Implementation

from app.aggregate.workload import Issue
from app.config import AppConfig


def _basic_auth_header(email: str, api_key: str) -> str:
    token = base64.b64encode(f"{email}:{api_key}".encode()).decode()
    return f"Basic {token}"


def _text_from_tool_result(result: Any) -> Any:
    structured = getattr(result, "structured_content", None)
    if structured is not None:
        return structured

    texts: list[str] = []
    for content in getattr(result, "content", None) or []:
        if getattr(content, "type", None) == "text" and getattr(content, "text", None):
            texts.append(content.text)
    if not texts:
        return None
    joined = "\n".join(texts)
    try:
        return json.loads(joined)
    except json.JSONDecodeError:
        return joined


def _extract_cloud_id(parsed: Any, site_url: str) -> str | None:
    host = urlparse(site_url).hostname or ""
    if isinstance(parsed, list):
        for item in parsed:
            if isinstance(item, dict):
                url = str(item.get("url") or item.get("siteUrl") or "")
                if host in url and isinstance(item.get("id"), str):
                    return item["id"]
                if host in url and isinstance(item.get("cloudId"), str):
                    return item["cloudId"]
    if isinstance(parsed, dict):
        resources = parsed.get("resources")
        if isinstance(resources, list):
            return _extract_cloud_id(resources, site_url)
        if isinstance(parsed.get("cloudId"), str):
            return parsed["cloudId"]
    return None


async def _resolve_cloud_id_from_site(site_url: str) -> str | None:
    try:
        async with httpx.AsyncClient(timeout=30.0) as client:
            res = await client.get(f"{site_url.rstrip('/')}/_edge/tenant_info")
            if res.status_code >= 400:
                return None
            data = res.json()
            cloud_id = data.get("cloudId")
            return cloud_id if isinstance(cloud_id, str) else None
    except Exception:
        return None


def _normalize_mcp_issues(payload: Any) -> tuple[list[Issue], bool, str | None]:
    if not payload:
        return [], False, None

    issues_raw: list[Any] = []
    total = 0
    start_at = 0
    max_results = 0
    next_page_token: str | None = None
    is_last: bool | None = None

    if isinstance(payload, list):
        issues_raw = payload
    elif isinstance(payload, dict):
        if isinstance(payload.get("issues"), list):
            issues_raw = payload["issues"]
        elif isinstance(payload.get("values"), list):
            issues_raw = payload["values"]
        total = int(payload.get("total") or 0)
        start_at = int(payload.get("startAt") or 0)
        max_results = int(payload.get("maxResults") or 0)
        token = payload.get("nextPageToken")
        if isinstance(token, str) and token:
            next_page_token = token
        if isinstance(payload.get("isLast"), bool):
            is_last = payload["isLast"]

    issues: list[Issue] = []
    for raw in issues_raw:
        if not isinstance(raw, dict):
            continue
        fields = raw.get("fields") if isinstance(raw.get("fields"), dict) else raw
        status = fields.get("status") if isinstance(fields.get("status"), dict) else {}
        status_category = (
            status.get("statusCategory")
            if isinstance(status.get("statusCategory"), dict)
            else {}
        )
        assignee = fields.get("assignee") if isinstance(fields.get("assignee"), dict) else None
        issues.append(
            Issue(
                key=str(raw.get("key") or fields.get("key") or ""),
                status_name=str(status.get("name") or "Unknown"),
                status_category=status_category.get("name")
                or status_category.get("key"),
                assignee_name=(assignee or {}).get("displayName")
                or (assignee or {}).get("name"),
            )
        )

    if is_last is True:
        return issues, False, None
    if next_page_token:
        return issues, True, next_page_token

    if total > 0:
        has_more = start_at + len(issues) < total
    else:
        has_more = max_results > 0 and len(issues) >= max_results
    return issues, has_more, None


async def fetch_issues_via_mcp(
    config: AppConfig,
) -> dict[str, Any]:
    """
    Returns either:
      { source: "mcp", issues: [...], tools: [...] }
    or
      { source: "mcp-unavailable", tools: [...], reason: "..." }
    """
    auth = _basic_auth_header(config.JIRA_EMAIL, config.JIRA_API_KEY)
    http_client = create_mcp_http_client(headers={"Authorization": auth})

    async with http_client:
        async with streamable_http_client(
            config.mcp_url_str, http_client=http_client
        ) as (read_stream, write_stream):
            async with ClientSession(
                read_stream,
                write_stream,
                client_info=Implementation(
                    name="jira-workload-worker",
                    version="0.1.0",
                ),
            ) as session:
                await session.initialize()
                listed = await session.list_tools()
                tools = [t.name for t in listed.tools]

                if "searchJiraIssuesUsingJql" not in tools:
                    return {
                        "source": "mcp-unavailable",
                        "tools": tools,
                        "reason": (
                            "MCP session has no searchJiraIssuesUsingJql tool. "
                            "Create an API token with Jira read/search scopes, and ensure "
                            "Rovo MCP API-token auth is enabled for your org."
                        ),
                    }

                cloud_id: str | None = None
                if "getAccessibleAtlassianResources" in tools:
                    resources = await session.call_tool(
                        "getAccessibleAtlassianResources",
                        arguments={},
                    )
                    parsed = _text_from_tool_result(resources)
                    cloud_id = _extract_cloud_id(parsed, config.site_str)
                if not cloud_id:
                    cloud_id = await _resolve_cloud_id_from_site(config.site_str)
                if not cloud_id:
                    return {
                        "source": "mcp-unavailable",
                        "tools": tools,
                        "reason": (
                            "Could not resolve Atlassian cloudId for MCP Jira tools. "
                            "Check JIRA_SITE."
                        ),
                    }

                jql = f"project = {config.JIRA_PROJECT_KEY} ORDER BY updated DESC"
                args: dict[str, Any] = {
                    "cloudId": cloud_id,
                    "jql": jql,
                    "maxResults": 100,
                    "fields": ["summary", "status", "assignee"],
                }

                all_issues: list[Issue] = []
                seen: set[str] = set()
                next_page_token: str | None = None
                start_at = 0

                for _ in range(20):
                    page_args = dict(args)
                    if next_page_token:
                        page_args["nextPageToken"] = next_page_token
                    elif start_at > 0:
                        page_args["startAt"] = start_at

                    result = await session.call_tool(
                        "searchJiraIssuesUsingJql",
                        arguments=page_args,
                    )
                    if getattr(result, "is_error", False):
                        err_text = _text_from_tool_result(result)
                        return {
                            "source": "mcp-unavailable",
                            "tools": tools,
                            "reason": (
                                "MCP searchJiraIssuesUsingJql failed: "
                                f"{str(err_text)[:300]}"
                            ),
                        }

                    payload = _text_from_tool_result(result)
                    page_issues, has_more, token = _normalize_mcp_issues(payload)
                    added = 0
                    for issue in page_issues:
                        if not issue.key or issue.key in seen:
                            continue
                        seen.add(issue.key)
                        all_issues.append(issue)
                        added += 1

                    if added == 0:
                        break
                    if token:
                        next_page_token = token
                        continue
                    if not has_more:
                        break
                    start_at += len(page_issues)

                return {
                    "source": "mcp",
                    "issues": all_issues,
                    "tools": tools,
                }
